firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteLocalStoreTest.java

```diff
@@ -398,19 +399,50 @@ public void testCanAutoCreateIndexes() {
     assertQueryReturned("coll/a", "coll/e", "coll/f");
   }
 
+  @Test
+  public void testCanAutoCreateIndexesWorksWithOrQuery() {
+    Query query = query("coll").filter(orFilters(filter("a", "==", 3), filter("b", "==", true)));
+    int targetId = allocateQuery(query);
+
+    setIndexAutoCreationEnabled(true);
+    setMinCollectionSizeToAutoCreateIndex(0);
+    setRelativeIndexReadCostPerDocument(2);
+
+    applyRemoteEvent(addedRemoteEvent(doc("coll/a", 10, map("b", true)), targetId));
+    applyRemoteEvent(addedRemoteEvent(doc("coll/b", 10, map("b", false)), targetId));
+    applyRemoteEvent(addedRemoteEvent(doc("coll/c", 10, map("a", 5, "b", false)), targetId));
+    applyRemoteEvent(addedRemoteEvent(doc("coll/d", 10, map("a", true)), targetId));
+    applyRemoteEvent(addedRemoteEvent(doc("coll/e", 10, map("a", 3, "b", true)), targetId));
+
+    // First time query runs without indexes.
+    // Based on current heuristic, collection document counts (5) > 2 * resultSize (2).
+    // Full matched index should be created.
+    executeQuery(query);
+    assertRemoteDocumentsRead(/* byKey= */ 0, /* byCollection= */ 2);
+    assertQueryReturned("coll/e", "coll/a");
+
+    backfillIndexes();
+
+    applyRemoteEvent(addedRemoteEvent(doc("coll/f", 20, map("a", 3, "b", false)), targetId));
+
+    executeQuery(query);
+    assertRemoteDocumentsRead(/* byKey= */ 2, /* byCollection= */ 1);
+    assertQueryReturned("coll/f", "coll/e", "coll/a");
+  }
+
   @Test
   public void testDoesNotAutoCreateIndexesForSmallCollections() {
-    Query query = query("coll").filter(filter("count", ">=", 3));
+    Query query = query("coll").filter(filter("foo", "==", 9)).filter(filter("count", ">=", 3));
```

Comment:
Could you explain the reason for this change? It doesn't seem to be any different from before given the name of the test, "testDoesNotAutoCreateIndexesForSmallCollections". That is, the collection is "small" in both cases. I ask because this seems like you're trying to add some test coverage which may warrant its own test method rather than extending this test method to cover two cases.

Reply:
`cover two cases` -> yes, this is what I try to do here.

The reason for that is I try to diversify the queries used in indexing tests. For other tests in this file, I tests queries on different types of fields. Then I released these tests are focus on one field, so I decided to change some of that to test on multiple fields.

Reply:
Ok that makes sense. For unit tests it's usually more readable to separate tests for distinct cases into distinct methods. For example, this test, by its name, is testing that indexes are _not_ created when the collection is too small. So the query should be as simple as possible so as not to distract from what is actually being testing. By making the query more complex it detracts from the readability of the test case because there is a lot of "noise". For example, looking at the documents that are added it takes some time to determine which of those documents would actually match the query. If, instead, the query just filtered on "foo=42" then it would be immediately obvious which documents match.

So consider breaking up this (and some other) tests into two different test methods: one for testing that indexes are _not_ created when the collection is too small and one for testing two filters (an equality and an inequality). If you'd like to do this but not right now then you can create a taskflow task as a reminder.

http://go/tott/649